File: apps/ios/Sources/LocalAgent/LocalAgentConfig.swift
import Foundation
import Security

enum LocalAgentProvider: String, CaseIterable, Codable {
    case anthropic = "anthropic"
    case openai = "openai"

    var displayName: String {
        switch self {
        case .anthropic: "Anthropic (Claude)"
        case .openai: "OpenAI (GPT)"
        }
    }

    var defaultModel: String {
        switch self {
        case .anthropic: "claude-sonnet-4-20250514"
        case .openai: "gpt-4o"
        }
    }

    var availableModels: [String] {
        switch self {
        case .anthropic:
            ["claude-sonnet-4-20250514", "claude-3-5-sonnet-20241022", "claude-3-5-haiku-20241022", "claude-3-haiku-20240307"]
        case .openai:
            ["gpt-4o", "gpt-4o-mini", "gpt-4-turbo", "o1-mini"]
        }
    }
}

@MainActor
final class LocalAgentConfig {
    static let shared = LocalAgentConfig()

    private static let keychainServicePrefix = "ai.openclaw.ios.localagent"
    private static let providerKey = "localAgent.provider"
    private static let modelKey = "localAgent.model"
    private static let enabledKey = "localAgent.enabled"
    private static let systemPromptKey = "localAgent.systemPrompt"
    private static let maxTokensKey = "localAgent.maxTokens"
    private static let customEndpointKey = "localAgent.customEndpoint"

    var isEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: Self.enabledKey) }
        set { UserDefaults.standard.set(newValue, forKey: Self.enabledKey) }
    }

    var provider: LocalAgentProvider {
        get {
            let raw = UserDefaults.standard.string(forKey: Self.providerKey) ?? ""
            return LocalAgentProvider(rawValue: raw) ?? .anthropic
        }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: Self.providerKey) }
    }

    var model: String {
        get {
            let stored = UserDefaults.standard.string(forKey: Self.modelKey) ?? ""
            return stored.isEmpty ? self.provider.defaultModel : stored
        }
        set { UserDefaults.standard.set(newValue, forKey: Self.modelKey) }
    }

    var systemPrompt: String {
        get {
            UserDefaults.standard.string(forKey: Self.systemPromptKey)
                ?? "You are a helpful AI assistant running locally on a mobile device. Be concise and helpful."
        }
        set { UserDefaults.standard.set(newValue, forKey: Self.systemPromptKey) }
    }

    var maxTokens: Int {
        get {
            let stored = UserDefaults.standard.integer(forKey: Self.maxTokensKey)
            return stored > 0 ? stored : 4096
        }
        set { UserDefaults.standard.set(newValue, forKey: Self.maxTokensKey) }
    }

    var customEndpoint: String {
        get { UserDefaults.standard.string(forKey: Self.customEndpointKey) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: Self.customEndpointKey) }
    }

    var hasAPIKey: Bool {
        let key = self.apiKey(for: self.provider)
        return !key.isEmpty
    }

    func apiKey(for provider: LocalAgentProvider) -> String {
        Self.readKeychain(account: "apiKey.\(provider.rawValue)") ?? ""
    }

    func setAPIKey(_ key: String, for provider: LocalAgentProvider) {
        if key.isEmpty {
            Self.deleteKeychain(account: "apiKey.\(provider.rawValue)")
        } else {
            Self.writeKeychain(account: "apiKey.\(provider.rawValue)", value: key)
        }
    }

    // MARK: - Keychain helpers

    private static func writeKeychain(account: String, value: String) {
        guard let data = value.data(using: .utf8) else { return }
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainServicePrefix,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
        var add = query
        add[kSecValueData as String] = data
        add[kSecAttrAccessible as String] = kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        SecItemAdd(add as CFDictionary, nil)
    }

    private static func readKeychain(account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainServicePrefix,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess, let data = result as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private static func deleteKeychain(account: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainServicePrefix,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

File: apps/android/app/src/main/java/ai/openclaw/android/agent/LocalAgentConfig.kt
package ai.openclaw.android.agent

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys

enum class AIProvider(val id: String, val displayName: String, val defaultModel: String) {
  ANTHROPIC("anthropic", "Anthropic (Claude)", "claude-sonnet-4-20250514"),
  OPENAI("openai", "OpenAI (GPT)", "gpt-4o");

  val availableModels: List<String>
    get() = when (this) {
      ANTHROPIC -> listOf(
        "claude-sonnet-4-20250514",
        "claude-3-5-sonnet-20241022",
        "claude-3-5-haiku-20241022",
        "claude-3-haiku-20240307",
      )
      OPENAI -> listOf("gpt-4o", "gpt-4o-mini", "gpt-4-turbo", "o1-mini")
    }

  companion object {
    fun fromId(id: String): AIProvider = entries.find { it.id == id } ?: ANTHROPIC
  }
}

class LocalAgentConfig(context: Context) {
  private val prefs: SharedPreferences = context.getSharedPreferences("local_agent", Context.MODE_PRIVATE)

  private val securePrefs: SharedPreferences by lazy {
    val masterKey = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
    EncryptedSharedPreferences.create(
      "local_agent_secure",
      masterKey,
      context,
      EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
      EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )
  }

  var isEnabled: Boolean
    get() = prefs.getBoolean("enabled", false)
    set(value) = prefs.edit().putBoolean("enabled", value).apply()

  var provider: AIProvider
    get() = AIProvider.fromId(prefs.getString("provider", "anthropic") ?: "anthropic")
    set(value) = prefs.edit().putString("provider", value.id).apply()

  var model: String
    get() = prefs.getString("model", null)?.takeIf { it.isNotEmpty() } ?: provider.defaultModel
    set(value) = prefs.edit().putString("model", value).apply()

  var systemPrompt: String
    get() = prefs.getString("systemPrompt", null)
      ?: "You are a helpful AI assistant running locally on a mobile device. Be concise and helpful."
    set(value) = prefs.edit().putString("systemPrompt", value).apply()

  var maxTokens: Int
    get() = prefs.getInt("maxTokens", 4096)
    set(value) = prefs.edit().putInt("maxTokens", value).apply()

  var customEndpoint: String
    get() = prefs.getString("customEndpoint", "") ?: ""
    set(value) = prefs.edit().putString("customEndpoint", value).apply()

  val hasApiKey: Boolean
    get() = getApiKey(provider).isNotEmpty()

  fun getApiKey(provider: AIProvider): String =
    securePrefs.getString("apiKey.${provider.id}", "") ?: ""

  fun setApiKey(key: String, provider: AIProvider) {
    securePrefs.edit().putString("apiKey.${provider.id}", key).apply()
  }
}
